File: src/test/clojure/com/brightcove/johnny/parts/host.clj
(ns com.brightcove.johnny.parts.host
  "Tests for host parsing."
  (:require [clojure.test :refer :all])
  (:import com.brightcove.johnny.http.Urls
           com.brightcove.johnny.parts.Host
           java.net.InetAddress
           com.google.common.net.InternetDomainName))

(deftest invariants
  (let [ip4 (InetAddress/getByName "4.4.4.4")
        ip4h (Host. ip4)]
    (is (.isIp ip4h))
    (is (= (.getIp ip4h) ip4)))
  (let [ip6 (InetAddress/getByName "::1")
        ip6h (Host. ip6)]
    (is (.isIp ip6h))
    (is (= (.getIp ip6h) ip6)))
  (let [dom (InternetDomainName/from "example.net")
        domh (Host. dom)]
    (is (.isDomain domh))
    (is (= (.getDomain domh) dom))))

(deftest which
  ;; Test component parsing in full-url-parsing context
  ;; to ensure that decoding happens.
  (let [p #(.getHost (Urls/parse (str "http://" %)))]
    (are [in out] (case out
                    :domain (is (.isDomain (p in)))
                    :ip (is (.isIp (p in)))
                    :inv (is (thrown? Exception (p in))))
         "www.example.com." :domain
         "localhost" :domain
         #_ (comment ;; JNUriParser rejects this
              "➡.ws" :domain)
         "xn--hgi.ws." :domain
         "www" :domain
         "www.." :inv
         ".localhost" :inv
         "256.com" :domain
         "192.168.1.115" :ip
         #_ (comment ;; Guava rejects this -- investigate
              "192.168.001.115" :ip)
         "[::1]" :ip
         "::1" :inv
         "[::1" :inv
         #_ (comment ;; TODO: zone IDs (Guava's InetAddresses lacks support)
              "[0:4::1%25eth0]" :ip
              "[0:4::1%eth0]" :inv))))

File: src/main/java/com/brightcove/johnny/parts/Host.java
package com.brightcove.johnny.parts;

import java.net.InetAddress;

import com.google.common.net.InternetDomainName;

import com.brightcove.johnny.http.Urls;

/**
 * The host component of a URL (e.g. "example.com" or an IP address.)
 *
 * Note that a hostname is not necessarily a domain name (e.g. IP addresses),
 * and a domain name is not necessarily a hostname (e.g. domain names with
 * underscores in them.)
 *
 * Delegates to {@link InternetDomainName} for domain names and
 * {@link InetAddress} for IP address literals.
 */
public class Host {

    private final InternetDomainName domain;
    private final InetAddress ip;

    /** Create a Host specified by a domain name. */
    public Host(InternetDomainName domain) {
        this.domain = domain;
        this.ip = null;
    }

    /** Create a Host specified by an IP address. */
    public Host(InetAddress ip) {
        this.domain = null;
        this.ip = ip;
    }

    /** Whether this host is specified by a domain name. */
    public boolean isDomain() { return domain != null; }

    /** Whether this host is specified by an IP address. */
    public boolean isIp() { return ip != null; }

    /** Return domain name if {@link #isDomain()} is true, else null */
    public InternetDomainName getDomain() {
        return domain;
    }

    /** Return IP address if {@link #isIp()} is true, else null */
    public InetAddress getIp() {
        return ip;
    }

    @Override
    public String toString() {
        return Urls.DEFAULT_HOST_ENCODER.unparse(this);
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Host)) {
            return false;
        }
        Host other = (Host) o;
        return isDomain() ? other.isDomain() && getDomain().equals(other.getDomain())
                          : getIp().equals(other.getIp());
    }

    @Override
    public int hashCode() {
        if (isDomain()) {
            return getDomain().hashCode();
        } else {
            return getIp().hashCode();
        }
    }
}
